### PR TITLE
Add Forwarding Preference=Datagram

### PR DESCRIPTION
**Pull request by afrind**
This was previously part of #333. It allows an object to be sent in a datagram f it fits, and put on a stream if it doesn't.

Closes #316

### CODE REVIEW

**Comment by afrind** [comment]
Linking some feedback raised in the original PR that has not yet been addressed:

https://github.com/moq-wg/moq-transport/pull/333#discussion_r1422925734

https://github.com/moq-wg/moq-transport/pull/333#discussion_r1422937755

**Review by suhasHere — APPROVED** [review]

File: draft-ietf-moq-transport.md [hunk 1]
```diff
@@ -676,6 +676,8 @@ MOQT Message {
 |------:|:---------------------------------------------------|
 | 0x0   | OBJECT_STREAM ({{object-message-formats}})         |
 |-------|----------------------------------------------------|
+| 0x1   | OBJECT_DATAGRAM_PREFERRED ({{object-message-formats}}) |
```
Comment (ianswett) [inline draft-ietf-moq-transport.md:679]: How about OBJECT_PREFER_DATAGRAM ?

File: draft-ietf-moq-transport.md [hunk 2]
```diff
@@ -917,6 +921,36 @@ receiver MUST close the session with a Protocol Violation.
 
 * Other fields: As described in {{canonical-object-fields}}.
 
+**Object Datagram Preferred Message**
+
+An `OBJECT_DATAGRAM_PREFERRED` message carries a single object in a datagram or
+a stream. There is no explicit length of the payload; it is determined by the
+length of the datagram or stream.  If this message appears on a stream, it MUST
+be the first message on a unidirectional stream.
```
Comment (ianswett) [inline draft-ietf-moq-transport.md:929]:
```suggestion
be the first and only message on a unidirectional stream.
```

File: draft-ietf-moq-transport.md [hunk 3]
```diff
@@ -917,6 +921,36 @@ receiver MUST close the session with a Protocol Violation.
...
+
+An Object received in an `OBJECT_DATAGRAM_PREFERRED` message has an `Object
+Forwarding Preference` = `Datagram`.
+
+To send an Object with `Object Forwarding Preference` = `Datagram`, determine
+the length of the fields and payload, and compare the length with the maximum
+datagram size of the session.  If the object size is less than or equal maximum
+datagram size, send the serialized data as a datagram.  Otherwise, open a
+stream, send the serialized data and terminate the stream.  An implementation
+MUST NOT send an Object with `Object Forwarding Preference` = `Datagram` on a
```
Comment (ianswett) [inline draft-ietf-moq-transport.md:939]: Are we going to enforce this MUST?  If not, I'd be inclined to say SHOULD NOT

**Review by ianswett — COMMENTED** [review]
Some minor comments, but LGTM

**Review by ianswett — APPROVED** [review]

File: draft-ietf-moq-transport.md [hunk 4]
```diff
@@ -931,6 +935,36 @@ the receiver MUST close the session with a Protocol Violation.
 
 * Other fields: As described in {{canonical-object-fields}}.
 
+**Object Datagram Preferred Message**
```
Comment (ianswett) [inline draft-ietf-moq-transport.md:938]:
```suggestion
**Object Prefer Datagram Message**
```

File: draft-ietf-moq-transport.md [hunk 5]
```diff
@@ -931,6 +935,36 @@ the receiver MUST close the session with a Protocol Violation.
...
+
+An Object received in an `OBJECT_PREFER_DATAGRAM` message has an `Object
+Forwarding Preference` = `Datagram`.
+
+To send an Object with `Object Forwarding Preference` = `Datagram`, determine
+the length of the fields and payload, and compare the length with the maximum
+datagram size of the session.  If the object size is less than or equal maximum
+datagram size, send the serialized data as a datagram.  Otherwise, open a
+stream, send the serialized data and terminate the stream.  An implementation
+MUST NOT send an Object with `Object Forwarding Preference` = `Datagram` on a
```
Comment (ianswett) [inline draft-ietf-moq-transport.md:953]:
```suggestion
SHOULD NOT send an Object with `Object Forwarding Preference` = `Datagram` on a
```

**Pull request merged**